Build a regression model.

In [1]:
# imports
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm
import warnings
warnings.filterwarnings("ignore")

# load data from part 1 and 2 into dataframes
city_bikes = pd.read_csv('citybike.csv')
fsq0_df = pd.read_csv('fsq_df_0.csv')
fsq1_df = pd.read_csv('fsq_df_1.csv')
fsq2_df = pd.read_csv('fsq_df_2.csv')
fsq3_df = pd.read_csv('fsq_df_3.csv')
fsq4_df = pd.read_csv('fsq_df_4.csv')
fsq5_df = pd.read_csv('fsq_df_5.csv')
yelp0_df = pd.read_csv('yelp_df_0.csv')
yelp1_df = pd.read_csv('yelp_df_1.csv')
yelp2_df = pd.read_csv('yelp_df_2.csv')
yelp3_df = pd.read_csv('yelp_df_3.csv')
yelp4_df = pd.read_csv('yelp_df_4.csv')
yelp5_df = pd.read_csv('yelp_df_5.csv')

In [67]:
# creating variable for the additional colums to be added
a = fsq0_df['rating'].mean()
b = fsq1_df['rating'].mean()
c = fsq2_df['rating'].mean()
d = fsq3_df['rating'].mean()
e = fsq4_df['rating'].mean()
f = fsq5_df['rating'].mean()


g = yelp0_df['rating'].mean()
h = yelp1_df['rating'].mean()
i = yelp2_df['rating'].mean()
j = yelp3_df['rating'].mean()
k = yelp4_df['rating'].mean()
l = yelp5_df['rating'].mean()


d0 = yelp0_df['distance'].mean()
d1 = yelp1_df['distance'].mean()
d2 = yelp2_df['distance'].mean()
d3 = yelp3_df['distance'].mean()
d4 = yelp4_df['distance'].mean()
d5 = yelp5_df['distance'].mean()

In [68]:
#Add the columns
bikes =city_bikes.copy()
bikes = bikes.drop('Unnamed: 0', axis = 1)
bikes = bikes.head(6).copy()
bikes['fsq_avg_rating(max = 10.0)'] = [a,b,c,d,e,f]
bikes['yelp_avg_rating(max = 5.0)'] = [g,h,i,j,k,l]
bikes['yelp_avg_distance'] = [d0,d1,d2,d3,d4,d5]

In [69]:
# deop unwanted columns
bikes = bikes.drop(['latitude', 'longitude', 'timestamp'], axis = 1)

In [59]:
bikes

empty_slots  free_bikes                    name  \
0            7           4         Clancy's Strand   
1           13           8  Saint Mary's Cathedral   
2           13           7    Island Road Junction   
3           11           9            Abbey Bridge   
4            6          10      O'Callaghan Strand   
5           12           7             Bedford Row   

   fsq_avg_rating(max = 10.0)  yelp_avg_rating(max = 5.0)  yelp_avg_distance  
0                    7.875000                       4.500         563.101144  
1                    7.833333                       4.525         527.862938  
2                    8.225000                       4.500         661.294203  
3                    8.092308                       4.500         668.383128  
4                    7.741176                       4.475         458.928216  
5                    7.640000                       4.450         258.529453

In [70]:
cols = [yelp0_df, yelp1_df, yelp2_df, yelp3_df, yelp4_df, yelp5_df]
yelp = pd.concat(cols, axis = 0, ignore_index = True)

In [71]:
yelp = yelp[['title', 'rating', 'review_count', 'distance']]
yelp.head()

title  rating  review_count    distance
0       Seafood     4.5           112   23.046412
1          Bars     4.5           158  442.238144
2     Irish Pub     5.0            16  252.343096
3  Asian Fusion     4.5            24  491.502548
4       Seafood     4.5            41  710.159997

In [23]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [40]:
categories = yelp['title'].tolist()
review = yelp['review_count'].tolist()
distance = yelp['distance'].tolist()
rating = yelp['rating'].tolist()


data = {
    'Category': categories,
    'rating' : rating,
    'review_count': review,
    'distance': distance
}

df = pd.DataFrame(data)


In [41]:
# Encode the categorical variable 'Category' using one-hot encoding
df_encoded = pd.get_dummies(df, columns=['Category'], drop_first=True)

# Split the data into features (X) and target variable (y)
X = df_encoded.drop('rating', axis=1)
y = df_encoded['rating']

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [42]:
# Create an instance of the LinearRegression model
model = LinearRegression()

# Fit the model to the training data
model.fit(X_train, y_train)

LinearRegression()

In [43]:
# Use the trained model to make predictions on the test set
y_pred = model.predict(X_test)


### Provide model output and an interpretation of the results. 

In [45]:
# Calculate evaluation metrics
mse = mean_squared_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred, squared=False)
r2 = r2_score(y_test, y_pred)

print('Mean Squared Error:', mse)
print('Root Mean Squared Error:', rmse)
print('R-squared:', r2)


Mean Squared Error: 0.017961927597805436
Root Mean Squared Error: 0.13402211607718123
R-squared: 0.6271686379698764


In [63]:
# Split into input (X) and output (y) variables
X = bikes[['yelp_avg_rating(max = 5.0)', 'yelp_avg_distance', 'yelp_avg_distance']]
y = bikes['empty_slots']

In [64]:
# Add constant column to input variables
X = sm.add_constant(X)

In [65]:
# Fit the OLS model
model = sm.OLS(y, X).fit()

In [66]:
# Retrieve the model summary
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:            empty_slots   R-squared:                       0.076
Model:                            OLS   Adj. R-squared:                 -0.540
Method:                 Least Squares   F-statistic:                    0.1233
Date:                Mon, 05 Jun 2023   Prob (F-statistic):              0.888
Time:                        01:57:03   Log-Likelihood:                -14.473
No. Observations:                   6   AIC:                             34.95
Df Residuals:                       3   BIC:                             34.32
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
const               

### Interpretation of the different aspects:

The R-squared value is 0.076, indicating that only 7.6% of the variance in the dependent variable (empty_slots) can be explained by the independent variables in the model. 

The adjusted R-squared value is -0.540, which suggests that the model may be overfitting or the chosen independent variables are not relevant in explaining the dependent variable.

Overall, based on the provided data, the regression model seems to have limited explanatory power, with low R-squared and adjusted R-squared values, non-significant coefficients, and a lack of statistical significance in the F-test.

# Stretch

How can you turn the regression model into a classification model?

### 
To turn the above regression model into a classification model, you would need to redefine the problem and modify the approach. Regression models are used to predict continuous numerical values, while classification models are used to predict discrete categories or classes.

Here are a few possible approaches to transforming the regression model into a classification model:

Define threshold values: You can define threshold values for the dependent variable (empty_slots) and convert it into a binary classification problem. For example, you can set a threshold such as empty_slots <= 5 as the positive class (e.g., low availability) and empty_slots > 5 as the negative class (e.g., high availability). Then, you would use the independent variables (e.g., yelp_avg_rating, yelp_avg_distance) to predict the class labels using classification algorithms like logistic regression, decision trees, or random forests.

Create categorical classes: Instead of using a binary classification approach, you can divide the range of the dependent variable (empty_slots) into multiple categories or classes. For example, you can create classes like "low availability," "medium availability," and "high availability" based on certain ranges of empty_slots values. Then, you would use the independent variables to predict the respective class labels using classification algorithms.

One-vs-All (OvA) approach: If you have multiple categories or classes for the dependent variable (e.g., different types of availability levels), you can use the one-vs-all approach. In this approach, you would train multiple binary classification models, each one comparing one class against the rest. For example, you can have one model to predict "low availability" versus "non-low availability," another model for "medium availability" versus "non-medium availability," and so on.

It's important to note that the specific approach to transforming the regression model into a classification model would depend on the nature of your data, the problem you are trying to solve, and the specific requirements of your classification task.
